### PR TITLE
[FLINK-37914][table] Add built-in OBJECT_UPDATE function

### PR DESCRIPTION
**Pull request by raminqaf**
## What is the purpose of the change

This pull request implements the `OBJECT_UPDATE` built-in function as part of [FLIP-520](https://cwiki.apache.org/confluence/display/FLINK/FLIP-520%3A+Simplify+StructuredType+handling): Simplify StructuredType handling. The function allows users to update existing fields in structured objects by providing key-value pairs, enabling mutation operations on structured types in both SQL and Table API without requiring custom UDFs.

## Brief change log

- Added `ObjectUpdateInputTypeStrategy` for validating input arguments (structured object + key-value pairs)
- Added `ObjectUpdateTypeStrategy` for inferring return types (same as input structured type)
- Implemented `ObjectUpdateFunction` runtime function for performing field updates
- Added `OBJECT_UPDATE` to `BuiltInFunctionDefinitions` with proper type inference strategies
- Added Table API expression support via `objectUpdate()` method on expressions
- Updated SQL and Table API documentation with examples and usage patterns
- Added comprehensive validation for field names, types, and compatibility checking
- Implemented proper error handling for non-existent fields and type mismatches

## Verifying this change

This change added tests and can be verified as follows:

- Added unit tests in `ObjectUpdateInputTypeStrategyTest` for input validation scenarios
- Added integration tests in `StructuredFunctionsITCase` for end-to-end SQL functionality

## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes (new built-in function)
- The serializers: no
- The runtime per-record code paths (performance sensitive): yes (new scalar function execution)
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
- The S3 file system connector: no

## Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? docs / JavaDocs (updated SQL functions documentation and comprehensive JavaDoc comments)

### CODE REVIEW

File: docs/data/sql_functions.yml [hunk 1]
```diff
@@ -1219,6 +1219,35 @@ valueconstruction:
   - table: NUMERIC.rows
     description: Creates a NUMERIC interval of rows (commonly used in window creation).
 
+valuemodification:
+  - sql: OBJECT_UPDATE(object, key, value [, key, value , ...])
+    table: OBJECT.objectUpdate(key, value [, key, value , ...])
+    description: |
+      Updates existing fields in a structured object by providing key-value pairs.
```
Comment (davidradl) [inline docs/data/sql_functions.yml:1226]: I am curious:
- I assume that arrays are not included in this 
- are there restrictions on the values, i.e. can they be maps, lists, objects or arrays. If there are restrictions , we should document them. If there are no restrictions we should include examples of more complex objects. 
- in the example I assume "com.example.User" is a "path". 
     - could dots in the field name clash with the path?
     - could you give an example or pointer to documentaiton around how to construct these paths in the documentation - examples around nested objects would be useful - including an array of objects where you want to update the value of the 3rd object.    
- can we set null values. 
- if the field is defined as not nullable - do we error if an attempt to put a null there occurs.

Comment (raminqaf) [reply docs/data/sql_functions.yml:1226]: - Arrays should be supported. In the background, the `OBJECT_UPDATE` uses a structured type, and [this test](https://github.com/raminqaf/flink/blob/e24c76cd611671ad893096bb0d74f89cdfa4dbdb/flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StructuredFunctionsITCase.java#L105-L105) evaluates its functionality.
- There should be no restrictions on the values. We only check if the type of the updated value matches the one in the structured type.
- For now, we decided to skip any validation on the field name format. The user is allowed to write `OBJECT_UPDATE(obj, "my.field.name", 14)`. This should be up to the user to pass valid field names. For the first iteration, this validation was not necessary.
- If I understand you correctly, you cannot update an `ARRAY` type with `OBJECT_UPDATE`. The `OBJECT_UPDATE` applies to `STRUCTURED_TYPE`.
- Yes, values can be null
- Yes, validation errors will be thrown if the field name is null:
	- Here is the [validation check](https://github.com/raminqaf/flink/blob/e24c76cd611671ad893096bb0d74f89cdfa4dbdb/flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java#L136-L136), where we see if the field is a non-null String literal.
	- [Tests](https://github.com/raminqaf/flink/blob/e24c76cd611671ad893096bb0d74f89cdfa4dbdb/flink-table/flink-table-common/src/test/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategyTest.java#L96-L96) backing this logic.

File: docs/data/sql_functions.yml [hunk 2]
```diff
@@ -1219,6 +1219,35 @@ valueconstruction:
...
+      and values are the new values for those fields. At least one key-value pair must be provided.
+      The total number of arguments must be odd (object + pairs of key-value arguments).
+      
+      The result type is the same structured type as the input, with the specified fields
+      updated to their new values.
+      
+      ```sql
+      -- Update the 'name' field of a user object
+      OBJECT_UPDATE(OBJECT_OF('com.example.User', 'name', 'Bob', 'age', 14), 'name', 'Alice')
+      -- Returns: User{name='Alice', age=14})
```
Comment (twalthr) [inline docs/data/sql_functions.yml]:
```suggestion
      -- Returns: User{name='Alice', age=14}
```

File: docs/data/sql_functions.yml [hunk 3]
```diff
@@ -1219,6 +1219,35 @@ valueconstruction:
...
+      
+      -- Update multiple fields
+      OBJECT_UPDATE(OBJECT_OF('com.example.User', 'name', 'Bob', 'age', 14), 'name', 'Alice', 'age', 30)
+      -- Returns: User{name='Alice', age=30})
```
Comment (twalthr) [inline docs/data/sql_functions.yml]:
```suggestion
      -- Returns: User{name='Alice', age=30}
```

File: flink-python/pyflink/table/expression.py [hunk 4]
```diff
@@ -2193,6 +2193,55 @@ def json_unquote(self) -> 'Expression':
...
+        :return: expression representing the updated structured type with modified
+                 field values
+        """
+        gateway = get_gateway()
+        ApiExpressionUtils = (
+            gateway.jvm.org.apache.flink.table.expressions.ApiExpressionUtils
+        )
+        exprs = [
+            ApiExpressionUtils.objectToExpression(_get_java_expression(e)) for e in kv
+        ]
```
Comment (twalthr) [inline flink-python/pyflink/table/expression.py]: Is this pattern used somewhere else? why not `_varargs_op`?

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 5]
```diff
@@ -2488,4 +2489,48 @@ public OutType percentile(InType percentage, InType frequency) {
...
+     *
+     * <p>The function expects alternating key-value pairs where keys are field names (non-null
+     * strings) and values are the new values for those fields. At least one key-value pair must be
+     * provided.
+     *
+     * <p>Example usage:
+     *
+     * <pre>{@code
+     * // Create a structured object representing a user
+     * User userObject = objectOf("com.example.User", "name", "Bob", "age", 25);
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java]: These examples are a bit misleading. They read as if `User userObject = objectOf` is Java code and the function returns a class instance `User`. The explanation in docs is better.

File: flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java [hunk 6]
```diff
@@ -2488,4 +2489,48 @@ public OutType percentile(InType percentage, InType frequency) {
...
+     *
+     * }</pre>
+     *
+     * <p>The result type is the same structured type as the input, with the specified fields
+     * updated to their new values.
+     *
+     * @param kv key-value pairs where even-indexed elements are field names (strings) and
+     *     odd-indexed elements are the new values for those fields
+     * @return expression representing a new structured object with updated field values
+     * @see org.apache.flink.table.functions.BuiltInFunctionDefinitions#OBJECT_UPDATE
```
Comment (twalthr) [inline flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/api/internal/BaseExpressions.java]: remove this reference? rather internal
```suggestion
     * @see org.apache.flink.table.functions.BuiltInFunctionDefinitions#OBJECT_UPDATE
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+import java.util.stream.Collectors;
+
+/**
+ * Input type strategy for the {@link BuiltInFunctionDefinitions#OBJECT_UPDATE} function.
+ *
+ * <p>This strategy validates the input arguments for updating existing fields in a structured type:
+ *
+ * <ul>
+ *   <li>Ensures the function has an odd number of arguments (at least 3)
+ *   <li>Validates the first argument is a non-null structured type
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java]:
```suggestion
 *   <li>Validates the first argument is a non-null structured type
```
doesn't need to be non-null. if null, null is returned. this is common for SQL functions.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+            final Map<String, LogicalType> structuredTypeAttributes,
+            final String keyNameToBeUpdated) {
+        final DataType argumentValueDataType = argumentDataTypes.get(pos);
+        final LogicalType argumentValueLogicalType = argumentValueDataType.getLogicalType();
+
+        final LogicalType expectedType = structuredTypeAttributes.get(keyNameToBeUpdated);
+        final DataType expectedDataType = DataTypes.of(expectedType);
+
+        // Validate that the updated value type matches the expected type
+        if (!argumentValueLogicalType
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java]: This check can be purely logical. So `LogicalTypeCasts#supportsImplicitCast` could be an option. But I'm wondering if we should remove this validation entirely. Otherwise a CHAR(3) could not be stored in a CHAR(2). Any mutation should be allowed and needs to be checked by a different entity (e.g. during equals).

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java [hunk 9]
```diff
@@ -0,0 +1,121 @@
...
+
+/**
+ * Type strategy for the {@link BuiltInFunctionDefinitions#OBJECT_UPDATE} function.
+ *
+ * <p>This strategy infers the return type for the OBJECT_UPDATE function by:
+ *
+ * <ul>
+ *   <li>Extracting the field definitions from the input structured type
+ *   <li>Resolving the class from the structured type
+ *   <li>
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java]:
```suggestion

```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java [hunk 10]
```diff
@@ -0,0 +1,121 @@
...
+public class ObjectUpdateTypeStrategy implements TypeStrategy {
+
+    private static Class<?> extractClass(
+            final CallContext callContext, final StructuredType structuredType) {
+        final DataTypeFactory dataTypeFactory = callContext.getDataTypeFactory();
+        final ClassLoader classLoader = dataTypeFactory.getClassLoader();
+        String className = structuredType.getClassName().orElseThrow(IllegalStateException::new);
+
+        return StructuredType.resolveClass(classLoader, className)
+                .orElseThrow(IllegalStateException::new);
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java]:
```suggestion
                .orElseThrow(IllegalStateException::new);
```
class is optional

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StructuredFunctionsITCase.java [hunk 11]
```diff
@@ -197,6 +202,99 @@ private static Stream<TestSetSpec> objectOfTestCases() {
...
+                        .andDataTypes(DataTypes.INT(), DataTypes.STRING())
+                        .withFunction(Type1.Type1Constructor.class)
+                        .withFunction(Type2.Type2Constructor.class)
+                        .withFunction(NestedType.NestedConstructor.class)
+                        // Test update all fields equality
+                        .testResult(
+                                call("Type1Constructor", $("f0"), $("f1"))
+                                        .objectUpdate("a", 16, "b", "Alice"),
+                                "OBJECT_UPDATE(OBJECT_OF('"
+                                        + Type1.class.getName()
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StructuredFunctionsITCase.java:218]: add a test where type does not exist in classpath

File: flink-python/pyflink/table/expression.py (hunk 4 above)
Comment (raminqaf) [reply flink-python/pyflink/table/expression.py]: I have found this pattern in the printf function:
https://github.com/raminqaf/flink/blob/3cf5551a9ad0dbbb5f1040d582e77343dfa23b90/flink-python/pyflink/table/expression.py#L1461-L1473

File: docs/data/sql_functions.yml [hunk 12]
```diff
@@ -1219,6 +1219,35 @@ valueconstruction:
...
+valuemodification:
+  - sql: OBJECT_UPDATE(object, key, value [, key, value , ...])
+    table: OBJECT.objectUpdate(key, value [, key, value , ...])
+    description: |
+      Updates existing fields in a structured object by providing key-value pairs.
+      
+      This function takes a structured object and updates specified fields with new values.
+      The keys must be string literals that correspond to existing fields in the structured type.
+      If a key does not exist in the input object, an exception will be thrown.
+      If the value type is not compatible with the corresponding structured field type,
```
Comment (twalthr) [inline docs/data/sql_functions.yml]: update this sentence everywhere

File: docs/data/sql_functions.yml [hunk 13]
```diff
@@ -1219,6 +1219,35 @@ valueconstruction:
...
+      The keys must be string literals that correspond to existing fields in the structured type.
+      If a key does not exist in the input object, an exception will be thrown.
+      If the value type is not compatible with the corresponding structured field type, 
+      an exception will also be thrown.
+      
+      The function expects alternating key-value pairs where keys are field names (non-null strings) 
+      and values are the new values for those fields. At least one key-value pair must be provided.
+      The total number of arguments must be odd (object + pairs of key-value arguments).
+      
+      The result type is the same structured type as the input, with the specified fields
```
Comment (twalthr) [inline docs/data/sql_functions.yml]:
```suggestion
      The result type is the same structured type class, with the specified fields
```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+ *
+ * <p>This strategy validates the input arguments for updating existing fields in a structured type:
+ *
+ * <ul>
+ *   <li>Ensures the function has an odd number of arguments (at least 3)
+ *   <li>Validates the first argument is a structured type
+ *   <li>Validates that key arguments are non-null string literals
+ *   <li>Ensures field names are not repeated in the key-value pairs
+ *   <li>Ensures field names are part of the structured type's attributes
+ *   <li>Ensures field values match the expected types defined in the structured type
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateInputTypeStrategy.java]:
```suggestion

```

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+import java.util.stream.IntStream;
+
+/**
+ * Type strategy for the {@link BuiltInFunctionDefinitions#OBJECT_UPDATE} function.
+ *
+ * <p>This strategy infers the return type for the OBJECT_UPDATE function by:
+ *
+ * <ul>
+ *   <li>Extracting the field definitions from the input structured type
+ *   <li>Resolving the class from the structured type
```
Comment (twalthr) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/types/inference/strategies/ObjectUpdateTypeStrategy.java]:
```suggestion

```

File: flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StructuredFunctionsITCase.java [hunk 16]
```diff
@@ -197,6 +202,101 @@ private static Stream<TestSetSpec> objectOfTestCases() {
...
+                        .testSqlResult(
+                                "OBJECT_UPDATE(OBJECT_OF('not.existing.clazz', 'a', 42, 'b', 'Bob'), 'b', 'Alice')",
+                                Row.of(42, "Alice"),
+                                DataTypes.STRUCTURED(
+                                        "not.existing.clazz",
+                                        DataTypes.FIELD("a", DataTypes.INT().notNull()),
+                                        DataTypes.FIELD("b", DataTypes.CHAR(5).notNull())))
+                        // Test update field to null
+                        .testResult(
+                                objectOf(Type1.class, "a", 42, "b", "Bob")
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/java/org/apache/flink/table/planner/functions/StructuredFunctionsITCase.java:282]: add a test where the object itself is null, e.g. `OBJECT_UPDATED(CAST(NULL AS STRUCTURED<>), ...)`

File: flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ObjectUpdateFunction.java [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+    private Map<String, Integer> fieldNameToRowPosIndex = new HashMap<>();
+
+    public ObjectUpdateFunction(SpecializedFunction.SpecializedContext context) {
+        super(BuiltInFunctionDefinitions.OBJECT_UPDATE, context);
+        this.fieldNameToRowPosIndex = createIndex(context);
+    }
+
+    public RowData eval(RowData rowData, Object... fieldNameAndValuePairs) {
+        GenericRowData updatedRow = (GenericRowData) rowData;
```
Comment (twalthr) [inline flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/scalar/ObjectUpdateFunction.java:51]: return null if rowData is null.

**Review by twalthr — COMMENTED** [review]
Last review comment. Almost there :-)
